### PR TITLE
Use external interfaces in oneMKL for mklcpu and mklgpu backends

### LINKED ISSUE #114: Investigate switching to external oneMKL DPC++ APIs for mklcpu/mklgpu backend
# Summary
This issues is related to Intel oneMKL symbols used in mklgpu backend. Currently internal symbols are used to call oneMKL routines. However, these symbols can chance between releases, which will cause build failures in oneMKL open source interfaces. 

# Problem statement
The internal symbols can change between releases and cause build failures. 

# Details
Switching to external Intel oneMKL APIs will improve stability of the repo between releases.

### PR DESCRIPTION
**Pull request by andrewtbarker**
# Description

Replaces calls to internal oneMKL functions with calls to the usual external interface. This should make the mklcpu and mklgpu backends more robust and predictable.

Fixes #114 

112 out of 3338 tests fail after these changes due to the underlying issue in #204. In fact these tests were never really passing because we were previously testing the reference Netlib BLAS against itself rather than testing the mklcpu backend. 

The tests are intended to check the `cblas_` interface to reference BLAS against DPC++ interfaces to oneMKL BLAS. Previously these used the same symbols, so we tested reference BLAS against itself. In most cases ( [example](https://github.com/oneapi-src/oneMKL/blob/722f6c51b528a2bc09bf549b1097f7d8fdc37826/src/blas/backends/mklcpu/mklcpu_level1.cxx#L76) ) we tested the `cblas_` interface against itself. In a few cases that use an increment parameter ( [example](https://github.com/oneapi-src/oneMKL/blob/722f6c51b528a2bc09bf549b1097f7d8fdc37826/src/blas/backends/mklcpu/mklcpu_level1.cxx#L29) ), we tested instead the fortran interface. These latter tests are exactly the ones that fail with the changes in this PR because negative values for the increment parameter are handled differently in the two cases.

Fixing the tests will require more serious changes to how the project as a whole handles dynamic libraries as documented in #204.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

[test.txt](https://github.com/oneapi-src/oneMKL/files/9014351/test.txt)

### CODE REVIEW

File: src/blas/backends/mkl_common/mkl_batch.cxx
Comment (mmeterel) [inline src/blas/backends/mkl_common/mkl_batch.cxx]: 2020-2021 ==> 2022

File: src/blas/backends/mkl_common/mkl_extensions.cxx
Comment (mmeterel) [inline src/blas/backends/mkl_common/mkl_extensions.cxx]: 2020-2021 => 2022

File: src/blas/backends/mkl_common/mkl_level1.cxx
Comment (mmeterel) [inline src/blas/backends/mkl_common/mkl_level1.cxx]: 2020-2021 => 2022

File: src/blas/backends/mkl_common/mkl_level2.cxx
Comment (mmeterel) [inline src/blas/backends/mkl_common/mkl_level2.cxx]: 2020-2021 => 2022

File: src/blas/backends/mkl_common/mkl_level3.cxx
Comment (mmeterel) [inline src/blas/backends/mkl_common/mkl_level3.cxx]: 2020-2021 => 2022

File: src/blas/backends/mklcpu/mklcpu_level3.cpp
Comment (mmeterel) [inline src/blas/backends/mklcpu/mklcpu_level3.cpp]: What is the concern here? For compile time, we rely on the user to specify the matching device and environment variable (ENABLE_MKLCPU_BACKEND). For runtime, we select the backend based on device type in queue. Are you referring to compile time case? Anyhow, we would need a global solution for that case.

File: src/blas/backends/mklgpu/mklgpu_level3.cpp
Comment (mmeterel) [inline src/blas/backends/mklgpu/mklgpu_level3.cpp]: same comment as before

File: src/blas/backends/mkl_common/mkl_batch.cxx
Comment (andrewtbarker) [reply src/blas/backends/mkl_common/mkl_batch.cxx]: Fixed, thanks.

File: src/blas/backends/mklcpu/mklcpu_level3.cpp
Comment (andrewtbarker) [reply src/blas/backends/mklcpu/mklcpu_level3.cpp]: Not sure what I was worried about, comment removed.

**Review by mmeterel — APPROVED** [review]
Thanks for the PR Andrew. Could you please add a note to #204 about the fact that, nrm2/scal/asum tests will start failing once we merge this PR? These failures needs to be fixed as part of #204.

**Comment by andrewtbarker** [comment]
@mkrainiuk Can you take a look at this PR please when you get a chance?

File: src/blas/backends/mkl_common/mkl_extensions.cxx
Comment (mkrainiuk) [inline src/blas/backends/mkl_common/mkl_extensions.cxx:26]: Does new DPC++ API require Intel oneMKL sycl library for MKLCPU backend? Because as far as I remember only LAPACK adds sycl library for MKLCPU backend, all other domains build with only 3 libraries.

Comment (mkrainiuk) [reply src/blas/backends/mkl_common/mkl_extensions.cxx:26]: Here is how sycl library is enabled in LAPACK MKLCPU backend https://github.com/oneapi-src/oneMKL/blob/3fe1f9bb897256d8c4d8fcb0de258b12c17c5a87/src/lapack/backends/mklcpu/CMakeLists.txt#L23

Comment (andrewtbarker) [reply src/blas/backends/mkl_common/mkl_extensions.cxx:26]: Good catch!

I've made similar changes in `src/blas/backends/mklcpu/CMakeLists.txt`. I'm testing now building with only mklcpu backend - this worked for me before probably because I always built with GPU backend enabled, even if I ran tests on CPU backend.

Comment (andrewtbarker) [reply src/blas/backends/mkl_common/mkl_extensions.cxx:26]: With change to `CMakeLists.txt`, a build with only mklcpu backend builds and the tests run.

**Comment by andrewtbarker** [comment]
Ping @mkrainiuk can you approve if you don't have additional concerns? Thanks!

**Review by mkrainiuk — APPROVED** [review]
Looks good to me, thank you!

**Pull request merged**
